src/ips.rs

```diff
@@ -0,0 +1,523 @@
+
+        for i in 0..agraph.len() {
+            for j in 0..agraph[i].len() {
+                let edge = Edge::new(
+                    IpAddr::from_str(nodes[i].ip.as_str()).unwrap(),
+                    IpAddr::from_str(nodes[j].ip.as_str()).unwrap(),
+                );
+                graph.insert(edge);
+            }
+        }
+        graph
+    }
+
+    fn degree_centrality_avg(&self, degrees: &HashMap<IpAddr, u32>) -> f64 {
+        let mut sum = 0.0;
+        for degree in degrees.values() {
+            sum += *degree as f64;
+        }
+        sum / degrees.len() as f64
+    }
+
+    fn determine_degrees_factors(&mut self, degrees: &HashMap<IpAddr, u32>) {
+        let min = *degrees
+            .iter()
+            .min_by(|a, b| a.1.cmp(b.1))
+            .map(|m| m.1)
+            .unwrap();
+        let max = *degrees
+            .iter()
+            .max_by(|a, b| a.1.cmp(b.1))
+            .map(|m| m.1)
+            .unwrap();
+
+        self.degree_factors = NormalizationFactors {
+            min: min as f64,
+            max: max as f64,
+        };
+    }
+
+    fn determine_eigenvalue_factors(&mut self, eigenvalues: &HashMap<IpAddr, f64>) {
+        let min = *eigenvalues
+            .iter()
+            .min_by(|a, b| a.1.partial_cmp(b.1).unwrap())
+            .map(|m| m.1)
+            .unwrap();
+        let max = *eigenvalues
+            .iter()
+            .max_by(|a, b| a.1.partial_cmp(b.1).unwrap())
+            .map(|m| m.1)
+            .unwrap();
+
+        self.eigenvector_factors = NormalizationFactors { min, max };
+    }
+
+    fn determine_betweenness_factors(&mut self, nodes: &[Node]) {
+        let min = nodes
+            .iter()
+            .min_by(|a, b| a.betweenness.partial_cmp(&b.betweenness).unwrap())
+            .map(|m| m.betweenness)
+            .unwrap();
+        let max = nodes
+            .iter()
+            .max_by(|a, b| a.betweenness.partial_cmp(&b.betweenness).unwrap())
+            .map(|m| m.betweenness)
+            .unwrap();
+
+        self.betweenness_factors = NormalizationFactors { min, max };
+    }
+
+    fn determine_closeness_factors(&mut self, nodes: &[Node]) {
+        let min = nodes
+            .iter()
+            .min_by(|a, b| a.closeness.partial_cmp(&b.closeness).unwrap())
+            .map(|m| m.closeness)
+            .unwrap();
+        let max = nodes
+            .iter()
+            .max_by(|a, b| a.closeness.partial_cmp(&b.closeness).unwrap())
+            .map(|m| m.closeness)
+            .unwrap();
+
+        self.closeness_factors = NormalizationFactors { min, max };
+    }
+
+    fn rate_node(&self, node: &Node, degree: u32, eigenvalue: f64) -> f64 {
+        // Calculate rating for node
+
+        // Rating is a combination of the following factors:
+        // 1. Degree
+        let mut rating = ((degree as f64 - self.degree_factors.min)
+            / (self.degree_factors.max - self.degree_factors.min)
```

Comment:
Done (assertion).

Reply:
Sure... but would it be better to skip updating the rating in that case? 

Otherwise, it would be funny to see this in the description:
> How do you know the network reached ideal levels after applying algorithm adjustments? 
> Well, you will know... the algorithm will crash itself if that ever happens

😃

Reply:
Omitting factors that could result in zero division.